File: tools/parity/dispatch/repo-clone-path.ts
// TS dispatch for the repo-clone-path parity module: maps the shared vector
// function names to the real `src/main/git/repo-clone-path.ts` exports so the
// harness compares the live TS reference against the Rust port.
//
// `deriveValidatedClonePath` reads `process.platform` and throws on invalid
// input; we run it on the (posix) harness host and wrap the result into a
// `{ clonePath }` / `{ error }` value to match the Rust `Result` image.

import {
  deriveValidatedClonePath,
  getClonePathComparisonKey
} from '../../../src/main/git/repo-clone-path'

export function dispatch(fn: string, input: unknown): unknown {
  switch (fn) {
    case 'deriveValidatedClonePath': {
      const { url, destination } = input as { url: string; destination: string }
      try {
        return { clonePath: deriveValidatedClonePath({ url, destination }) }
      } catch (error) {
        return { error: (error as Error).message }
      }
    }
    case 'getClonePathComparisonKey': {
      const { clonePath } = input as { clonePath: string }
      return getClonePathComparisonKey(clonePath)
    }
    default:
      throw new Error(`unknown function ${fn}`)
  }
}

File: rust/crates/orca-parity/src/modules/repo_clone_path.rs
//! Parity dispatch for `orca_git::repo_clone_path` vs
//! `src/main/git/repo-clone-path.ts`.
//!
//! `deriveValidatedClonePath` throws on invalid input and returns the clone path
//! string on success; the Rust port returns `Result<String, String>` with the
//! same messages. Both are shaped into `{ clonePath }` / `{ error }` so the JSON
//! images are equal. `getClonePathComparisonKey` returns a plain string.

use orca_core::cross_platform_path::PathFlavor;
use orca_git::repo_clone_path::{derive_validated_clone_path, get_clone_path_comparison_key};
use serde_json::{json, Value};

/// The TS `deriveValidatedClonePath` reads `process.platform`; mirror the host so
/// the Rust port and the live TS agree. The goldens assume a posix host.
fn host_path_flavor() -> PathFlavor {
    if std::env::consts::OS == "windows" {
        PathFlavor::Windows
    } else {
        PathFlavor::Posix
    }
}

pub fn dispatch(function: &str, input: &Value) -> Value {
    match function {
        "deriveValidatedClonePath" => {
            let url = input.get("url").and_then(Value::as_str).unwrap_or("");
            let destination = input.get("destination").and_then(Value::as_str).unwrap_or("");
            match derive_validated_clone_path(url, destination, host_path_flavor()) {
                Ok(clone_path) => json!({ "clonePath": clone_path }),
                Err(message) => json!({ "error": message }),
            }
        }
        "getClonePathComparisonKey" => {
            let clone_path = input.get("clonePath").and_then(Value::as_str).unwrap_or("");
            Value::String(get_clone_path_comparison_key(clone_path))
        }
        other => json!({ "__parity_error__": format!("unknown function {other}") }),
    }
}
